### PR TITLE
Merge Clever and Dense dicts

### PR DESCRIPTION
**Pull request by joaquimg**

### CODE REVIEW

File: src/Utilities/CleverDicts.jl [hunk 1]
```diff
@@ -33,7 +37,11 @@ Note that querying a `LinearIndex` immediately after deleting a key via
 Store an item `val` using `add_item(c::CleverDict, val)`. `add_item` returns a
 key corresponding to the stored item.
 
-Overload the functions `index_to_key` and `key_to_index` to enable mappings
+It is possible to initialize the CleverDict as `CleverDict{K, V}(n)` so that
+`n` elements are pre-allocated and can be efficiently added even out of order
+as long as theirkey hashes are between ´1´ and `n`.
```
Comment (blegat) [inline src/Utilities/CleverDicts.jl:42]: Missing space after their

File: src/Utilities/CleverDicts.jl [hunk 2]
```diff
@@ -33,7 +37,11 @@ Note that querying a `LinearIndex` immediately after deleting a key via
...
+
+Overload the functions `index_to_key` and `key_to_index` to enable vectorpings
```
Comment (blegat) [inline src/Utilities/CleverDicts.jl:44]: "vectorpings"?

File: src/Utilities/CleverDicts.jl [hunk 3]
```diff
@@ -46,11 +54,28 @@ index_to_key(::Type{MyKey}, i::Int) = MyKey(i)
 key_to_index(key::MyKey) = key.x
 ```
 """
-mutable struct CleverDict{K, V} <: AbstractDict{K, V}
+mutable struct CleverDict{K, V, F, I} <: AbstractDict{K, V}
```
Comment (blegat) [inline src/Utilities/CleverDicts.jl:57]: `F<:Function`

File: src/Utilities/CleverDicts.jl [hunk 4]
```diff
@@ -63,68 +88,119 @@ function index_to_key end
...
+        if !haskey(c, key)
+            throw(KeyError(key))
+        end
+        return c.vector[c.hash(key)::Int]
+    else
+        return c.dict[key]
+    end
+end
+function Base.setindex!(c::CleverDict{K, V}, value, key)::V where {K, V}
+    h = c.hash(key)::Int
```
Comment (blegat) [inline src/Utilities/CleverDicts.jl:151]: `Int64`

File: src/Utilities/CleverDicts.jl (hunk 1 above)
Comment (odow) [reply src/Utilities/CleverDicts.jl:42]: Also ´1´ -> `1`

File: src/Utilities/CleverDicts.jl (hunk 3 above)
Comment (odow) [reply src/Utilities/CleverDicts.jl:57]: Is this going to break the downstream solvers?

File: src/Utilities/CleverDicts.jl [hunk 5]
```diff
@@ -63,68 +88,119 @@ function index_to_key end
 """
     key_to_index(key::K)
 
-Map `key` to an integer valued index, assuming that there have been no
+vector `key` to an integer valued index, assuming that there have been no
```
Comment (odow) [inline src/Utilities/CleverDicts.jl:91]: ?

File: src/Utilities/CleverDicts.jl (hunk 3 above)
Comment (joaquimg) [reply src/Utilities/CleverDicts.jl:57]: Haven't tested, but it is possible.
After writing the entire thing I think that it ok to use the function overloading for now.

Comment (joaquimg) [reply src/Utilities/CleverDicts.jl]: Yet another non-breaking possibility is to have AbstractCleverDict with 4 params and the Clever Dict with 2, no super good though...

Comment (odow) [reply src/Utilities/CleverDicts.jl]: What's the hit of just making them `::Function`? Or making `index_to_key` dependent on the `K-V` pair: `key_to_index(::CleverDict{K, V}, x::K)::Int` and `index_to_key(::CleverDict{K, V}, v::Int)::K`,

Comment (joaquimg) [reply src/Utilities/CleverDicts.jl]: just `::Function` is not good for inference, and the type wont be concrete.

I like:
> making index_to_key dependent on the K-V pair: key_to_index(::CleverDict{K, V}, x::K)::Int and index_to_key(::CleverDict{K, V}, v::Int)::K

Comment (odow) [reply src/Utilities/CleverDicts.jl]: If you go the second route, we will need a deprecated fallback.

**Comment by joaquimg** [comment]
@odow it turns out that the way it currently is does not break GLPK, CPLEX, Gurobi nor Xpress.
Any other solver I should test?

They might have a small inference penalty because their structs will be referencing an abstract type.
A simple straight forward PR fixes this non-breaking problem.

If this one gets merged I can quickly PR an optimized `_reverse_dict` for `CleverDict`.

**Comment by blegat** [comment]
> They might have a small inference penalty because their structs will be referencing an abstract type.

This is because we added type arguments to CleverDict right? So adding it make it concrete?

**Comment by joaquimg** [comment]
> This is because we added type arguments to CleverDict right? So adding it make it concrete?

Yes, the PRs only have to add 2 parameters to the type definition of the CleverDict in the optimizer structs.

**Comment by blegat** [comment]
Any objection to merge?

**Comment by joaquimg** [comment]
It is good for me!

**Pull request merged**